File: swift-bridge/Sources/ModelIOBridge/Animation.swift
import Foundation
import ModelIO
import simd

private func mdl_double_matrix_to_float_array(_ matrix: matrix_double4x4) -> [Float] {
    [
        Float(matrix.columns.0.x), Float(matrix.columns.0.y), Float(matrix.columns.0.z), Float(matrix.columns.0.w),
        Float(matrix.columns.1.x), Float(matrix.columns.1.y), Float(matrix.columns.1.z), Float(matrix.columns.1.w),
        Float(matrix.columns.2.x), Float(matrix.columns.2.y), Float(matrix.columns.2.z), Float(matrix.columns.2.w),
        Float(matrix.columns.3.x), Float(matrix.columns.3.y), Float(matrix.columns.3.z), Float(matrix.columns.3.w),
    ]
}

private func mdl_float_array_to_double_matrix(_ values: UnsafePointer<Float>?) -> matrix_double4x4 {
    let matrix = mdl_matrix_from_array(values)
    return matrix_double4x4(columns: (
        SIMD4<Double>(Double(matrix.columns.0.x), Double(matrix.columns.0.y), Double(matrix.columns.0.z), Double(matrix.columns.0.w)),
        SIMD4<Double>(Double(matrix.columns.1.x), Double(matrix.columns.1.y), Double(matrix.columns.1.z), Double(matrix.columns.1.w)),
        SIMD4<Double>(Double(matrix.columns.2.x), Double(matrix.columns.2.y), Double(matrix.columns.2.z), Double(matrix.columns.2.w)),
        SIMD4<Double>(Double(matrix.columns.3.x), Double(matrix.columns.3.y), Double(matrix.columns.3.z), Double(matrix.columns.3.w))
    ))
}

private func mdl_packed_joint_animation_info(_ animation: MDLPackedJointAnimation) -> [String: Any] {
    [
        "name": animation.name,
        "path": animation.path,
        "joint_paths": animation.jointPaths,
        "joint_count": animation.jointPaths.count,
    ]
}

private func mdl_animation_bind_component_info(_ component: MDLAnimationBindComponent) -> [String: Any] {
    [
        "has_skeleton": component.skeleton != nil,
        "has_joint_animation": component.jointAnimation != nil,
        "joint_paths": component.jointPaths as Any,
        "geometry_bind_transform": mdl_double_matrix_to_float_array(component.geometryBindTransform),
    ]
}

@_cdecl("mdl_packed_joint_animation_new")
public func mdl_packed_joint_animation_new(
    _ name: UnsafePointer<CChar>?,
    _ jointPaths: UnsafePointer<UnsafePointer<CChar>?>?,
    _ jointPathCount: UInt64,
    _ outAnimation: UnsafeMutablePointer<UnsafeMutableRawPointer?>?,
    _ outError: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?
) -> Int32 {
    mdl_run(outError) {
        guard let name, let outAnimation else {
            throw ModelIOBridgeError.invalidArgument("missing animation name or output pointer")
        }
        outAnimation.pointee = mdl_retain(
            MDLPackedJointAnimation(
                name: String(cString: name),
                jointPaths: mdl_string_array(jointPaths, count: jointPathCount)
            )
        )
    }
}

@_cdecl("mdl_packed_joint_animation_info_json")
public func mdl_packed_joint_animation_info_json(_ handle: UnsafeMutableRawPointer?) -> UnsafeMutablePointer<CChar>? {
    guard let animation = mdl_borrow_object(handle) as? MDLPackedJointAnimation else { return nil }
    return mdl_string(mdl_json_string(from: mdl_packed_joint_animation_info(animation)) ?? "{}")
}

@_cdecl("mdl_packed_joint_animation_translations")
public func mdl_packed_joint_animation_translations(_ handle: UnsafeMutableRawPointer?) -> UnsafeMutableRawPointer? {
    guard let animation = mdl_borrow_object(handle) as? MDLPackedJointAnimation else { return nil }
    return mdl_retain(animation.translations)
}

@_cdecl("mdl_packed_joint_animation_rotations")
public func mdl_packed_joint_animation_rotations(_ handle: UnsafeMutableRawPointer?) -> UnsafeMutableRawPointer? {
    guard let animation = mdl_borrow_object(handle) as? MDLPackedJointAnimation else { return nil }
    return mdl_retain(animation.rotations)
}

@_cdecl("mdl_packed_joint_animation_scales")
public func mdl_packed_joint_animation_scales(_ handle: UnsafeMutableRawPointer?) -> UnsafeMutableRawPointer? {
    guard let animation = mdl_borrow_object(handle) as? MDLPackedJointAnimation else { return nil }
    return mdl_retain(animation.scales)
}

@_cdecl("mdl_animation_bind_component_new")
public func mdl_animation_bind_component_new(
    _ outComponent: UnsafeMutablePointer<UnsafeMutableRawPointer?>?,
    _ outError: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?
) -> Int32 {
    mdl_run(outError) {
        guard let outComponent else {
            throw ModelIOBridgeError.invalidArgument("missing output animation bind component pointer")
        }
        outComponent.pointee = mdl_retain(MDLAnimationBindComponent())
    }
}

@_cdecl("mdl_animation_bind_component_info_json")
public func mdl_animation_bind_component_info_json(_ handle: UnsafeMutableRawPointer?) -> UnsafeMutablePointer<CChar>? {
    guard let component = mdl_borrow_object(handle) as? MDLAnimationBindComponent else { return nil }
    return mdl_string(mdl_json_string(from: mdl_animation_bind_component_info(component)) ?? "{}")
}

@_cdecl("mdl_animation_bind_component_set_skeleton")
public func mdl_animation_bind_component_set_skeleton(_ handle: UnsafeMutableRawPointer?, _ skeletonHandle: UnsafeMutableRawPointer?) {
    guard let component = mdl_borrow_object(handle) as? MDLAnimationBindComponent else { return }
    component.skeleton = mdl_borrow_object(skeletonHandle) as? MDLSkeleton
}

@_cdecl("mdl_animation_bind_component_skeleton")
public func mdl_animation_bind_component_skeleton(_ handle: UnsafeMutableRawPointer?) -> UnsafeMutableRawPointer? {
    guard let component = mdl_borrow_object(handle) as? MDLAnimationBindComponent,
          let skeleton = component.skeleton
    else {
        return nil
    }
    return mdl_retain(skeleton)
}

@_cdecl("mdl_animation_bind_component_set_packed_joint_animation")
public func mdl_animation_bind_component_set_packed_joint_animation(_ handle: UnsafeMutableRawPointer?, _ animationHandle: UnsafeMutableRawPointer?) {
    guard let component = mdl_borrow_object(handle) as? MDLAnimationBindComponent else { return }
    component.jointAnimation = mdl_borrow_object(animationHandle) as? MDLPackedJointAnimation
}

@_cdecl("mdl_animation_bind_component_packed_joint_animation")
public func mdl_animation_bind_component_packed_joint_animation(_ handle: UnsafeMutableRawPointer?) -> UnsafeMutableRawPointer? {
    guard let component = mdl_borrow_object(handle) as? MDLAnimationBindComponent,
          let animation = component.jointAnimation as? MDLPackedJointAnimation
    else {
        return nil
    }
    return mdl_retain(animation)
}

@_cdecl("mdl_animation_bind_component_set_joint_paths")
public func mdl_animation_bind_component_set_joint_paths(
    _ handle: UnsafeMutableRawPointer?,
    _ jointPaths: UnsafePointer<UnsafePointer<CChar>?>?,
    _ jointPathCount: UInt64
) {
    guard let component = mdl_borrow_object(handle) as? MDLAnimationBindComponent else { return }
    component.jointPaths = mdl_string_array(jointPaths, count: jointPathCount)
}

@_cdecl("mdl_animation_bind_component_set_geometry_bind_transform")
public func mdl_animation_bind_component_set_geometry_bind_transform(_ handle: UnsafeMutableRawPointer?, _ values: UnsafePointer<Float>?) {
    guard let component = mdl_borrow_object(handle) as? MDLAnimationBindComponent else { return }
    component.geometryBindTransform = mdl_float_array_to_double_matrix(values)
}

File: src/animation.rs
use std::ffi::CString;
use std::ptr;

use crate::animated_value_types::{AnimatedQuaternionArray, AnimatedVector3Array};
use crate::error::Result;
use crate::ffi;
use crate::handle::ObjectHandle;
use crate::object::Object;
use crate::protocols::{Component, JointAnimation};
use crate::skeleton::Skeleton;
use crate::types::{AnimationBindComponentInfo, PackedJointAnimationInfo};
use crate::util::{c_string, parse_json, required_handle};

fn c_string_vec(values: &[&str]) -> Result<(Vec<CString>, Vec<*const i8>)> {
    let c_strings = values
        .iter()
        .map(|value| c_string(value))
        .collect::<Result<Vec<_>>>()?;
    let raw = c_strings.iter().map(|value| value.as_ptr()).collect();
    Ok((c_strings, raw))
}

#[derive(Debug, Clone)]
/// Wraps the corresponding Model I/O packed joint animation counterpart.
pub struct PackedJointAnimation {
    handle: ObjectHandle,
}

impl JointAnimation for PackedJointAnimation {}

impl PackedJointAnimation {
    /// Builds this wrapper from the retained handle of the wrapped Model I/O packed joint animation counterpart.
    pub(crate) fn from_handle(handle: ObjectHandle) -> Self {
        Self { handle }
    }

    /// Wraps the corresponding Model I/O initializer for the wrapped Model I/O packed joint animation counterpart.
    pub fn new(name: &str, joint_paths: &[&str]) -> Result<Self> {
        let name = c_string(name)?;
        let (_joint_paths, raw_joint_paths) = c_string_vec(joint_paths)?;
        let mut out_animation = ptr::null_mut();
        let mut out_error = ptr::null_mut();
        // SAFETY: The unsafe operation is valid in this context.
        let status = unsafe {
            ffi::mdl_packed_joint_animation_new(
                name.as_ptr(),
                raw_joint_paths.as_ptr(),
                raw_joint_paths.len() as u64,
                &mut out_animation,
                &mut out_error,
            )
        };
        crate::util::status_result(status, out_error)?;
        Ok(Self::from_handle(required_handle(
            out_animation,
            "MDLPackedJointAnimation",
        )?))
    }

    /// Calls the corresponding Model I/O method on the wrapped Model I/O packed joint animation counterpart.
    pub fn info(&self) -> Result<PackedJointAnimationInfo> {
        parse_json(
            // SAFETY: ObjectHandle wraps a valid opaque pointer from Swift; FFI function accepts it safely.
            unsafe { ffi::mdl_packed_joint_animation_info_json(self.handle.as_ptr()) },
            "MDLPackedJointAnimation",
        )
    }

    /// Calls the corresponding Model I/O method on the wrapped Model I/O packed joint animation counterpart.
    pub fn translations(&self) -> Result<AnimatedVector3Array> {
        // SAFETY: ObjectHandle wraps a valid opaque pointer from Swift; FFI function accepts it safely.
        let ptr = unsafe { ffi::mdl_packed_joint_animation_translations(self.handle.as_ptr()) };
        Ok(AnimatedVector3Array::from_handle(required_handle(
            ptr,
            "MDLPackedJointAnimation translations",
        )?))
    }

    /// Calls the corresponding Model I/O method on the wrapped Model I/O packed joint animation counterpart.
    pub fn rotations(&self) -> Result<AnimatedQuaternionArray> {
        // SAFETY: ObjectHandle wraps a valid opaque pointer from Swift; FFI function accepts it safely.
        let ptr = unsafe { ffi::mdl_packed_joint_animation_rotations(self.handle.as_ptr()) };
        Ok(AnimatedQuaternionArray::from_handle(required_handle(
            ptr,
            "MDLPackedJointAnimation rotations",
        )?))
    }

    /// Calls the corresponding Model I/O method on the wrapped Model I/O packed joint animation counterpart.
    pub fn scales(&self) -> Result<AnimatedVector3Array> {
        // SAFETY: ObjectHandle wraps a valid opaque pointer from Swift; FFI function accepts it safely.
        let ptr = unsafe { ffi::mdl_packed_joint_animation_scales(self.handle.as_ptr()) };
        Ok(AnimatedVector3Array::from_handle(required_handle(
            ptr,
            "MDLPackedJointAnimation scales",
        )?))
    }

    #[must_use]
    /// Calls the corresponding Model I/O method on the wrapped Model I/O packed joint animation counterpart.
    pub fn as_object(&self) -> Object {
        Object::from_handle(self.handle.clone())
    }
}

#[derive(Debug, Clone)]
/// Wraps the corresponding Model I/O animation bind component counterpart.
pub struct AnimationBindComponent {
    handle: ObjectHandle,
}

impl Component for AnimationBindComponent {}

impl AnimationBindComponent {
    /// Builds this wrapper from the retained handle of the wrapped Model I/O animation bind component counterpart.
    pub(crate) fn from_handle(handle: ObjectHandle) -> Self {
        Self { handle }
    }

    /// Wraps the corresponding Model I/O initializer for the wrapped Model I/O animation bind component counterpart.
    pub fn new() -> Result<Self> {
        let mut out_component = ptr::null_mut();
        let mut out_error = ptr::null_mut();
        let status =
            // SAFETY: Output pointers are initialized and managed; FFI function is called safely.
            unsafe { ffi::mdl_animation_bind_component_new(&mut out_component, &mut out_error) };
        crate::util::status_result(status, out_error)?;
        Ok(Self::from_handle(required_handle(
            out_component,
            "MDLAnimationBindComponent",
        )?))
    }

    /// Calls the corresponding Model I/O method on the wrapped Model I/O animation bind component counterpart.
    pub fn info(&self) -> Result<AnimationBindComponentInfo> {
        parse_json(
            // SAFETY: ObjectHandle wraps a valid opaque pointer from Swift; FFI function accepts it safely.
            unsafe { ffi::mdl_animation_bind_component_info_json(self.handle.as_ptr()) },
            "MDLAnimationBindComponent",
        )
    }

    /// Calls the corresponding Model I/O method on the wrapped Model I/O animation bind component counterpart.
    pub fn set_skeleton(&self, skeleton: &Skeleton) {
        // SAFETY: The unsafe operation is valid in this context.
        unsafe {
            ffi::mdl_animation_bind_component_set_skeleton(
                self.handle.as_ptr(),
                skeleton.handle.as_ptr(),
            );
        };
    }

    #[must_use]
    /// Calls the corresponding Model I/O method on the wrapped Model I/O animation bind component counterpart.
    pub fn skeleton(&self) -> Option<Skeleton> {
        // SAFETY: ObjectHandle wraps a valid opaque pointer from Swift; FFI function accepts it safely.
        let ptr = unsafe { ffi::mdl_animation_bind_component_skeleton(self.handle.as_ptr()) };
        // SAFETY: The unsafe operation is valid in this context.
        unsafe { ObjectHandle::from_retained_ptr(ptr) }.map(Skeleton::from_handle)
    }

    /// Calls the corresponding Model I/O method on the wrapped Model I/O animation bind component counterpart.
    pub fn set_packed_joint_animation(&self, animation: &PackedJointAnimation) {
        // SAFETY: The unsafe operation is valid in this context.
        unsafe {
            ffi::mdl_animation_bind_component_set_packed_joint_animation(
                self.handle.as_ptr(),
                animation.handle.as_ptr(),
            );
        };
    }

    #[must_use]
    /// Calls the corresponding Model I/O method on the wrapped Model I/O animation bind component counterpart.
    pub fn packed_joint_animation(&self) -> Option<PackedJointAnimation> {
        // SAFETY: The unsafe operation is valid in this context.
        let ptr = unsafe {
            ffi::mdl_animation_bind_component_packed_joint_animation(self.handle.as_ptr())
        };
        // SAFETY: The unsafe operation is valid in this context.
        unsafe { ObjectHandle::from_retained_ptr(ptr) }.map(PackedJointAnimation::from_handle)
    }

    /// Calls the corresponding Model I/O method on the wrapped Model I/O animation bind component counterpart.
    pub fn set_joint_paths(&self, joint_paths: &[&str]) -> Result<()> {
        let (_joint_paths, raw_joint_paths) = c_string_vec(joint_paths)?;
        // SAFETY: The unsafe operation is valid in this context.
        unsafe {
            ffi::mdl_animation_bind_component_set_joint_paths(
                self.handle.as_ptr(),
                raw_joint_paths.as_ptr(),
                raw_joint_paths.len() as u64,
            );
        };
        Ok(())
    }

    /// Calls the corresponding Model I/O method on the wrapped Model I/O animation bind component counterpart.
    pub fn set_geometry_bind_transform(&self, matrix: [f32; 16]) {
        // SAFETY: The unsafe operation is valid in this context.
        unsafe {
            ffi::mdl_animation_bind_component_set_geometry_bind_transform(
                self.handle.as_ptr(),
                matrix.as_ptr(),
            );
        };
    }
}
